### ISSUE TITLE
Not starting on lock screen after reboot.

### ISSUE THREAD

**Issue by DobriniaMerk** [dessalines/thumb-key#1036]
### Steps to reproduce

1. Disable all keyboards except thumb-key, including the stock one.
2. Set screen lock to password (not PIN).
3. Reboot the phone.
4. Try to enter the password.

### Expected behavior

thumb-key should start.

### Actual behavior

thumb-key does not show up to enter the password, and you get a soft-bricked device.

### version of the program

3.3.5

### Android version

Android 14

### Device

Tecno Pova 5

### Other details

Issue I found while trying to unlock my device: https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/2445

About unlocking:
>TL;DR: If you have no trusted ADB RSA fingerprints, no custom recovery, and no USB OTG support, you're screwed.

### Acknowledgements

- [X] I have written a short but informative title.
- [X] I have updated the app to **[the latest version](https://github.com/dessalines/thumb-key/releases/latest)**.
- [X] I have searched the existing issues and this is a new one, **NOT** a  duplicate or related to another open issue.
- [X] This is not a question or a discussion, in which case I should have  gone to [lemmy.ml/c/thumbkey](https://lemmy.ml/c/thumbkey)
- [X] This is a **single** bug report, in case of multiple bugs I will open  a separate issue for each one  (they can always link to each other if related)
- [ ] I have admitted that I am a clown by having checked this box, as I have not read these acknowledgements. 🤡
- [X] I have filled out all of the requested information in this form.

**Comment by DobriniaMerk** [comment]
Actually, I haven't updated to the latest version because it's hard to do on a locked device.

**Comment by dessalines** [comment]
That issue only seems applicable to android 10. I'm also a little scared to test this since it could brick any device.

**Comment by LucaMkIII** [comment]
I've noticed a pop-up message when I turn on Thumb-Key as an input method that mentions this (I'm on a sort-of stock Android 12 - I'm not sure if the pop-up is on stock Android or not). Checking for this pop-up could be an easy (and safe) first test.

- AnySoftKeyboard also shows the pop-up.
- Fossify Keyboard (https://github.com/FossifyOrg/Keyboard) does *not* give me this pop-up.

Here's a screen recording showing each.
https://github.com/user-attachments/assets/24f9e8a1-cd81-4a33-b937-f172cc0aa717

**Comment by SpacePirates6** [comment]
I want to preference this with I have very little experience on Android 1

But on newer android, if the apk isn't a system app then it cannot load without the device being unlocked the first time, If I remember correctly this is a function of the user space being encrypted before first unlock, you can achieve a similar soft lock if you forcibly remove gboard on using root or adb, as there is no expectation for a user to do such action they do not have a failsafe.

**Comment by mruncreative** [comment]
This may be near impossible to do as others have explained above.

**Comment by dafiish** [comment]
I soft-bricked my lineageos 20 device this way, too. I solved the problem by plugging in a hardware keyboard using a USB hub.

**Comment by alerque** [comment]
Is this still an issue—and if so is it really an issue with *this* keyboard? None of the Android devices I have allow me to have anything other than the system default (AOSP) keyboard available on the lock screen on first boot. Not Thumb-Key, not Google's proprietary keyboard, not any of the other third party ones I have. This is a system restriction and nothing to do with Thumb-Key.

Is there actually something here wrong with or that can be changed on our side?

**Comment by DobriniaMerk** [comment]
Just checked it on my device. It only loads the stock keyboard, so the problem might be nonexistent already. However, the popup behavior that LucaMkIII showed is still there. It would be strange if those popups mean nothing now but still exist.

**Comment by alerque** [comment]
> Just checked it on my device. It only loads the stock keyboard, so the problem might be nonexistent already. However, the popup behavior that LucaMkIII showed is still there. It would be strange if those popups mean nothing now but still exist.

The screen recording from @LucaMkIII clearly shows exactly the expected Android behavior: it clearly says that even though the keyboard is enabled that it is not a system keyboard and will NOT be able to run when the device first starts before the first unlock. It also shows this for other keyboards they select except the GBoard one, that has special access because their ROM has some parts of that keyboard registered by the system ROM, not just an external app.

I'm pretty sure there is nothing to do here. Android categorically does not allow 3rd party (non-system) apps to run before unlocking devices, which means the device unlock has to work with the system keyboard. That keyboard is always enabled by the system on first run ever if the user preferences switch to a different input device directly after. Nothing Thumb-Key does can change this (for better or worse).

I'm going to close this issue because I don't think there is actually an issue we can resolve. At best this discussion can educate users on what to expect from Android.

**Comment by DobriniaMerk** [comment]
> It also shows this for other keyboards they select except the GBoard one, that has special access because their ROM has some parts of that keyboard registered by the system ROM, not just an external app.

That is not exactly true. Fossyfy Keyboard, despite not being a system keyboard, clearly shows warning popup only about data collection, but not about starting before device unlock unlike every other non-system ones.

**Comment by alerque** [comment]
I just installed the Fossify keyboard to test and to my surprise not only load it not give that warning, but it actually did load at boot to unlock the device. I wonder how they achieved that. I did not that the unlock screen did not have the languages I setup so perhaps there is some special mode that can be set on a layout that is known to work for all allowed symbols or something like that.

**Comment by alerque** [comment]
> Fossyfy Keyboard, despite not being a system keyboard, clearly shows warning popup only about data collection, but not about starting before device unlock unlike every other non-system ones.

It seems there *is* a way to make keyboards explicitly compatible with device unlocking, but it's a little bit involved.

First off, the tail is a bit twisted because it turns out the Fossify Keyboard project is just a rebrand of [this project](https://github.com/SimpleMobileTools/Simple-Keyboard/pull/183). The PR which first introduced direct boot awareness af a feature was [this one](https://github.com/SimpleMobileTools/Simple-Keyboard/pull/183), with follow ups [here](https://github.com/SimpleMobileTools/Simple-Keyboard/pull/192) and [here](https://github.com/SimpleMobileTools/Simple-Keyboard/pull/195).

At first blush it looks like we need some direct_boot aware default settings (e.g. a known-compatible layout, etc.) and some other adaptations.

Given that people probably would **not be able to have their customized layouts available at unlock time** since those may be languages or have miss-matched symbol sets with what is required to be an unlocker, is this really something people still want?

**Comment by ismi-ismi** [comment]
>  is this really something people still want?

yes of course. i have been getting soft-locked on my phone because this is the only keyboard i have, the first time it happened i ended up factory reset the phone. the second time, i found [scrcpy OTG](https://github.com/Genymobile/scrcpy/blob/master/doc/otg.md) where i can connect my phone to my laptop via USB and use the laptop keyboard to unlock my phone.

I do wish the implementation to make this keyboard appear at boot as a feature.
EDIT: Apologies for the aggression, I'm panicking and frustrated (when i made the original comment) that i almost brick my phone. Change my wording to make it less aggressive

**Comment by dessalines** [comment]
Did you read the comments above? This is likely impossible for us to do.

> But on newer android, if the apk isn't a system app then it cannot load without the device being unlocked the first time, If I remember correctly this is a function of the user space being encrypted before first unlock, you can achieve a similar soft lock if you forcibly remove gboard on using root or adb, as there is no expectation for a user to do such action they do not have a failsafe.

**Comment by ismi-ismi** [comment]
Oh, I thought it's possible because it seemed that people alluding that fossify keyboard able to do it. I apologize if it's likely impossible

**Comment by alerque** [comment]
@dessalines I thought this was going to be impossible too, but I did verify (on my encrypted devices no less) that it is possible now, and that Fossify successfully does it. I started looking through the code to figure out how and it looks like it will not be trivial. You can't just fire up the keyboard with all the user's preferences (e.g. layout), you have to have a special mode that works when no settings are available. As such even if we accomplish this (which I'm still looking into) it may or may not address everybody's use case because one default layout may not even allow them to enter their password, much less in the preferred familiar way.

**Comment by alerque** [comment]
> I DO NOT WANT MY PHONE TO GET SOFT-LOCKED AGAIN!

Dude stop yelling at people doing FOSS work, that's rude and entitled and (at least in my case) has exactly the opposite effect you want, it demotivates people working on other people's issues.

In the mean time per my previous message, even if we get this feature working using the relatively new Android APIs for it, it likely will not solve your issue because you would still not having anything other than a default (likely English based) layout of some kind, not your preferred layout and customizations.

To not lock yourself out of your phone your root decryption password **MUST** be something you can type on the default system keyboard language and layout without accessing the keyboard settings.

**Comment by dessalines** [comment]
For sure. If someone wanted to work on this I wouldn't be opposed, but I'm not testing it as it could soft brick my device.

It seems like newer android versions use a pin or swipe by default, which uses its own numeric input.

**Comment by alerque** [comment]
> I'm not testing it as it could soft brick my device.

I can test this because I have at least one devices I am prepared to hard reset anyway. But this really isn't quite as big a deal as people are making it out to be. Some people mistakenly hard reset their device to get out of a software keyboard pickle with their encrypted root, but that is actually not necessary. By default Android has USB HID devices support enabled, so all you need to do is plug in a keyboard or send keyboard events over ADB to enter the password. People just don't know to do that ... but it will get you out of almost any bind in relation to software keyboards.

**Comment by ismi-ismi** [comment]
> Dude stop yelling at people doing FOSS work, that's rude and entitled and (at least in my case) has exactly the opposite effect you want, it demotivates people working on other people's issues.

Yeah, i'm sorry. I'm panicking and frustrated (when i made the original comment) that i almost brick my phone.

**Comment by LucaMkIII** [comment]
> First off, the tail is a bit twisted because it turns out the Fossify Keyboard project is just a rebrand of [this project](https://github.com/SimpleMobileTools/Simple-Keyboard/pull/183). The PR which first introduced direct boot awareness af a feature was [this one](https://github.com/SimpleMobileTools/Simple-Keyboard/pull/183), with follow ups [here](https://github.com/SimpleMobileTools/Simple-Keyboard/pull/192) and [here](https://github.com/SimpleMobileTools/Simple-Keyboard/pull/195).

Yeah, the Fossify suite of apps was forked from Simple Mobile Tools by one of the former devs when SMT was sold to ZipoApps.

In case it helps anyone looking into this - Fossify Keyboard had [an issue](https://github.com/FossifyOrg/Keyboard/issues/59) where it crashed on start up. It was apparently fixed by [this commit](https://github.com/FossifyOrg/Keyboard/commit/da2280a523102271339dc7ad1643e1690b171ad9).

**Comment by k26pl** [comment]
As far as i know (tho i am not sure), apps that are marked as directBootAware get 2 directories for user data:
/data/user/<uid>/<packageid> - credential encrypted storage, the default directory, encrypted with user pin
/data/user_de/<n>/<packageid> - device encrypted storage
the key for 2nd one is managed by tee or something similiar, so it is available before user enters pin first time. If i am not mistaken, it should be possible to also store required data like layout here, so it can be accessed while phone is booted.  
https://developer.android.com/privacy-and-security/direct-boot

**Comment by alerque** [comment]
Thanks for the extra pointers @k26pl. I keep meaning to work on this but it keeps getting pushed off. If you get a chance to actually work an an implementation I could maybe collaborate. If what you said about storage was true it might make sense to copy a minimal set of settings from the user profile to the system data, e.g. keep the chosen layout(s) in sync, but not copy the users YAML customizations (which might have sensitive data), color scheme, etc.
